core/src/test/scala/caliban/relay/ConnectionSpec.scala

```diff
@@ -69,6 +69,69 @@ object ConnectionSpec extends DefaultRunnableSpec {
           )
       )
     },
+    test("it correctly slices the list when before is set") {
+      assertTrue(
+        ItemConnection
+          .fromList(
+            List(Item("a"), Item("b"), Item("c")),
+            Pagination(cursor = PaginationCursor.Before(Base64Cursor(2)), count = PaginationCount.First(3))
+          )
+          .edges == List(
+          ItemEdge(Base64Cursor(0), Item("a")),
+          ItemEdge(Base64Cursor(1), Item("b"))
+        )
+      )
+    },
+    test("it correctly calculates hasNextPage") {
+      val list = List(Item("a"), Item("b"), Item("c"))
+
+      val calcHasNextPage = (cursor: PaginationCursor[Base64Cursor], count: PaginationCount) =>
+        ItemConnection
+          .fromList(
+            list,
+            Pagination(cursor = cursor, count = count)
+          )
+          .pageInfo
+          .hasNextPage
+
+      // when first is set
+      assert(calcHasNextPage(PaginationCursor.NoCursor, PaginationCount.First(2)))(isTrue) &&
+      assert(calcHasNextPage(PaginationCursor.NoCursor, PaginationCount.First(3)))(isFalse) &&
+      assert(calcHasNextPage(PaginationCursor.After(Base64Cursor(0)), PaginationCount.First(1)))(isTrue) &&
+      assert(calcHasNextPage(PaginationCursor.After(Base64Cursor(1)), PaginationCount.First(1)))(isFalse) &&
+      // when before is set
+      assert(calcHasNextPage(PaginationCursor.Before(Base64Cursor(2)), PaginationCount.First(1)))(isTrue) &&
+      assert(calcHasNextPage(PaginationCursor.Before(Base64Cursor(2)), PaginationCount.First(2)))(isFalse)
```

Comment:
I think you're missing some && to combine the tests. Same below.

Reply:
ah ha! Good catch. Fixed at 93e8cf4a0f31c3eccbffbde3dfffcd770f9f2624